Automattic/vip
# FORNO-1704: SQL Import - Improve handling of compressed files

## Description

This PR aims to address some inconsistencies around how compressed files are handled during SQL imports. 

1. Currently only uncompressed and `gz` files can be successfully imported to a VIP environment using `vip import sql`, but we don't validate extensions in the CLI.
1. Currently running `vip import validate-sql` on a compressed file will return normal validation errors like `Error:  SQL Error: DROP TABLE was not found.` even if the uncompressed file is valid.
2. Currently only uncompressed files can be imported using `vip dev-env import sql`; compressed files are attempted to be validated and encounter the same issue as described in 2 above.

This PR addresses these issues by:
1. Adding validation to only permit `.sql` and `.gz` extensions for both `vip import sql` and `vip dev-env import sql`
3. Adding support for `.gz` archives in `vip dev-env import sql`, extracting the archive to a temp directory before importing the resulting SQL file as normal
4. Returning a useful error on `vip import validate-sql` to explain that compressed files can't be validated

## Steps to Test

1. Download the test files below which contain this SQL:

```
DROP TABLE IF EXISTS `wp_a8c_cron_control_jobs`;

CREATE TABLE `wp_a8c_cron_control_jobs` (
  `ID` bigint(20) UNSIGNED NOT NULL,
  `timestamp` bigint(20) UNSIGNED NOT NULL,
  `action` varchar(255) NOT NULL,
  `action_hashed` varchar(32) NOT NULL,
  `instance` varchar(32) NOT NULL,
  `args` longtext NOT NULL,
  `schedule` varchar(255) DEFAULT NULL,
  `interval` int(10) UNSIGNED DEFAULT 0,
  `status` varchar(32) NOT NULL DEFAULT 'pending',
  `created` datetime NOT NULL,
  `last_modified` datetime NOT NULL
) ENGINE=InnoDB DEFAULT CHARSET=utf8;
```

`.sql`: https://transfer.sh/5L7MiLwtQH/test.sql
`.sql.gz`: https://transfer.sh/1t712ht7JL/test.sql.gz
`.zip`: https://transfer.sh/KJOzR1paS9/test.zip

2. Check out PR.
3. Run `npm run build`

Validation flow:

4. Run `node ./dist/bin/vip-import-validate-sql.js test.zip`
5. Run `node ./dist/bin/vip-import-validate-sql.js test.sql.gz`
6. Verify the command exits and an error is shown
7. Run `node ./dist/bin/vip-import-validate-sql.js test.sql`
8. Verify the sql file validates

Dev Env flow:

9. Run `vip dev-env start --slug <your slug>` to start your local development environment
1. Run `node ./dist/bin/vip-dev-env-import-sql.js test.zip  --slug <your slug>`
2. Verify the command fails with an error about the extension being unsupported.
1. Run `node ./dist/bin/vip-dev-env-import-sql.js test.sql.gz --slug <your slug>`
1. Run `node ./dist/bin/vip-dev-env-import-sql.js test.sql --slug <your slug>`
1. Verify both commands succeed and the file is imported

Import flow:

15. Run `node ./dist/bin/vip-import-sql.js test.zip @<id>.production`
2. Verify the command fails with an error about the extension being unsupported.
1. Run `node ./dist/bin/vip-import-sql.js test.sql @<id>.production` (don't confirm)
3. Run `node ./dist/bin/vip-import-sql.js test.sql.gz @<id>.production` (don't confirm)
